qcore/geo.py

```diff
@@ -1086,15 +1086,164 @@ def oriented_bounding_planes(
     return bounding_planes
 
 
+def closest_points_between_line_segments(
+    p1: np.ndarray, p2: np.ndarray, q1: np.ndarray, q2: np.ndarray
+) -> Tuple[np.ndarray, np.ndarray]:
+    """Find the closest pair of points between two line segments in 3D.
+
+    Given four points p1, p2, q1, q2, defining line segments l = <p1, p2> and m
+    = <q1, q2>, find the closest points between the line segments l and m.
+
+    Parameters
+    ----------
+    p1 : np.ndarray
+        First point on l.
+    p2 : np.ndarray
+        Second point on l.
+    q1 : np.ndarray
+        First point on m.
+    q2 : np.ndarray
+        Second point on m.
+
+    Returns
+    -------
+    (np.ndarray, np.ndarray)
+        The point pair (p, q) with p in l and q in m that minimizes ||p - q||.
+        If l and m are parallel, this pair is not uniquely defined.
+
+    Examples
+    --------
+    >>> closest_points_between_line_segments(np.array([0, 0, 0]),
+                                             np.array([1, 0, 0]),
+                                             np.array([0.5, 1, 1]),
+                                             np.array([0.5, -1, -1]))
+    (array([0.5, 0, 0]), array([0.5, 0, 1]))
+    """
+    # Suppsoe that l and m have points
+    #   p_s = p1 + s (p2 - p1), and
+    #   q_t = p3 + t (p4 - p3).
+    # Then, the values s and t minimising the distance
+    # between line segments has the vector p_s - q_t orthogonal to both p2 -
+    # p1 and p4 - p3. By expressing the conditions
+    #
+    #            (p_s - q_t) * (p2 - p1) = 0, and
+    #            (p_s - q_t) * (p4 - p3) = 0
+    #
+    # as a linear system with unknowns s and t, we can solve for s & t and
+    # get the closest points. This results in the following system:
+    #
+    #     ⎛s ⎞
+    #   A ⎜  ⎟ =  b,
+    #     ⎝-t⎠
+    #
+    # where
+    #             ⎛p  - p ⎞
+    #             ⎜ 2    1⎟ ⎛p  - p  q  - q ⎞
```

Comment:
can they be just p2 p1, instead of using two lines?

Reply:
Yeah, that's just the ascii mathematics formatter at work there. Easy fix.